### ISSUE TITLE
oneflow.layers.batch_normalization

### ISSUE THREAD

**Issue by kanagi2587** [Oneflow-Inc/oneflow#3929]
这个算子如何才能和torch.nn.BatchNorm系列算子对应上，尤其是axis这一参数，在torch的bn中是没有的，不知道如何才能对应上。
另外在对二维数据测试的过程中：

@flow.global_function()
def test(inputs:tp.Numpy.Placeholder((2,4),dtype=flow.float64)
)->tp.Numpy:
    inputs = flow.layers.batch_normalization(inputs=inputs,axis=1)
    return inputs
            

a=torch.tensor([[1.0,3.0,3.0,7.0],[1.0,4.0,3.0,4.0]])
bn=torch.nn.BatchNorm1d(4)
print(bn(a))
x=np.array([[1.0,3.0,3.0,7.0],[1.0,4.0,3.0,4.0]]).astype(np.float)
out=test(x)
print(out)
输出为
tensor([[ 0.0000, -1.0000,  0.0000,  1.0000],
        [ 0.0000,  1.0000,  0.0000, -1.0000]],
       grad_fn=<NativeBatchNormBackward>)
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]]
这里为什么都是0

在我的项目中大量应用torch中的bn算子，迫切需要知道如何修改oneflow中的bn算子（尤其是axis等）才能对应上。

**Comment by guo-ran** [comment]
1、上面结果是0的原因是没有调用下面的代码初始化模型
 ```
check_point = flow.train.CheckPoint()
check_point.init()
```
2、
axis 一般是C所在的维度，应该设1
3、
和torch对应的话，需要注意momentum 参数和torch中意义不同，torch momentum设0.1, oneflow需要设0.9

**Comment by kanagi2587** [comment]
> 1、上面结果是0的原因是没有调用下面的代码初始化模型

请问oneflow中的bn机制与pytorch中的有什么不同么，因为就算参数对应上了，对相同的张量进行测试，发现bn之后的数据大相径庭

**Comment by guo-ran** [comment]
batch norm在预测和训练时的行为不一致，预测时默认不根据输入x计算mean和variance，而是直接使用模型中的mean和variance参与计算。 训练时会根据输入x计算mean和variance，并相应更新模型。

上面的脚本改成下面这样，即训练模式，应该就是你想要的结果。

```

@flow.global_function(type="train", function_config=flow.FunctionConfig())
def test(inputs:tp.Numpy.Placeholder((2,4),dtype=flow.float64)
    )->tp.Numpy:
  inputs = flow.layers.batch_normalization(inputs=inputs,axis=1,momentum=0.9,epsilon=1e-5)
  flow.optimizer.SGD(flow.optimizer.PiecewiseConstantScheduler([], [0.001]), momentum=0).minimize(flow.math.reduce_sum(inputs))
  return inputs

check_point = flow.train.CheckPoint()
check_point.init()
x=np.array([[1.0,3.0,3.0,7.0],[1.0,4.0,3.0,4.0]]).astype(np.float)
out=test(x)
print(out)

**Comment by kanagi2587** [comment]
> batch norm在预测和训练时的行为不一致，预测时默认不根据输入x计算mean和variance，而是直接使用模型中的mean和variance参与计算。 训练时会根据输入x计算mean和variance，并相应更新模型。
> 
> 上面的脚本改成下面这样，即训练模式，应该就是你想要的结果。
> 
> ```
> 
> @flow.global_function(type="train", function_config=flow.FunctionConfig())
> def test(inputs:tp.Numpy.Placeholder((2,4),dtype=flow.float64)
>     )->tp.Numpy:
>   inputs = flow.layers.batch_normalization(inputs=inputs,axis=1,momentum=0.9,epsilon=1e-5)
>   flow.optimizer.SGD(flow.optimizer.PiecewiseConstantScheduler([], [0.001]), momentum=0).minimize(flow.math.reduce_sum(inputs))
>   return inputs
> 
> check_point = flow.train.CheckPoint()
> check_point.init()
> x=np.array([[1.0,3.0,3.0,7.0],[1.0,4.0,3.0,4.0]]).astype(np.float)
> out=test(x)
> print(out)
> ```

请问这个batch_normalization是否适用于从torch.nn.BatchNorm1d 到torch.nn.BatchNorm3d 一系列的替换，因为在排查网络中出现的数据量级的变化。  想知道用于BatchNorm3d 的话是否需要特殊的修改

**Comment by guo-ran** [comment]
1d 2d 3d都是通用的，下面是我构建的几个测试用例，oneflow和pytorch结果差不多：

### 2d
```
import oneflow as flow
import oneflow.typing as tp
import numpy as np
import torch

@flow.global_function(type="train", function_config=flow.FunctionConfig())
def test(inputs:tp.Numpy.Placeholder((4,4,4,5),dtype=flow.float)
    )->tp.Numpy:
  inputs = flow.layers.batch_normalization(inputs=inputs,axis=1,momentum=0.9,epsilon=1e-5)
  flow.optimizer.SGD(flow.optimizer.PiecewiseConstantScheduler([], [0.001]), momentum=0).minimize(flow.math.reduce_sum(inputs))
  return inputs

check_point = flow.train.CheckPoint()
check_point.init()
x=np.random.random((4, 4, 4, 5)).astype(np.float32)
out=test(x)
print("of_out", out.flatten()[0:10])
a=torch.tensor(x)
bn=torch.nn.BatchNorm2d(4)
torch_out = bn(a)
print("torch_out", torch_out.detach().numpy().flatten()[0:10])
```

### 3d
```
import oneflow as flow
import oneflow.typing as tp
import numpy as np
import torch

@flow.global_function(type="train", function_config=flow.FunctionConfig())
def test(inputs:tp.Numpy.Placeholder((5, 4,4,4,5),dtype=flow.float)
    )->tp.Numpy:
  inputs = flow.layers.batch_normalization(inputs=inputs,axis=1,momentum=0.9,epsilon=1e-5)
  flow.optimizer.SGD(flow.optimizer.PiecewiseConstantScheduler([], [0.001]), momentum=0).minimize(flow.math.reduce_sum(inputs))
  return inputs

check_point = flow.train.CheckPoint()
check_point.init()
x=np.random.random((5, 4, 4, 4, 5)).astype(np.float32)
out=test(x)
print("of_out", out.flatten()[0:10])
a=torch.tensor(x)
bn=torch.nn.BatchNorm3d(4)
torch_out = bn(a)
print("torch_out", torch_out.detach().numpy().flatten()[0:10])
